spmd/tensor/device_mesh.py

```diff
@@ -351,45 +402,72 @@ def all_reduce(
         op: ReduceOp = ReduceOp.SUM,  # type: ignore
         mesh_dim: int = 0,
     ):
+        """
+        all_reduce the tensor on each rank on a device mesh dimension, and
+        return an output tensor on each rank after all_reduce.
+
+        Args:
+            input (torch.Tensor): tensor to be all_reduced on each rank.
+            op (:class:`torch.distributed.distributed_c10d.ReduceOp, optional):
+                the reduction op of all_reduce (i.e. ReduceOp.SUM)
+            mesh_dim (int, optional): indicate which mesh dimension we want
+                to reduce on.
+
+        Returns:
+            A :class:`torch.Tensor` object
+        """
         dim_group = self._dim_groups[mesh_dim]
         # CommTensor does not change eager mode behavior. During tracing, it
         # makes sure communication result is properly waited before subsequent
         # read operations.
-        return all_reduce(
-            CommTensor(tensor.contiguous()), op=op, group=dim_group
-        )
+        if not tensor.is_contiguous():
+            tensor = CommTensor(tensor.contiguous())
+        else:
+            tensor = CommTensor(tensor.clone())
```

Comment:
Curious why we need to call `.clone()` instead of `tensor = CommTensor(tensor)`?

Reply:
we actually need to make the receiving tensors be a separate receiving tensor rather than itself, otherwise collective might write into it and the result becomes wrong. Actually we can just call clone in this case. let me update